| Organization | Course | Exercise | Semester | Professor | T.A. | 
| --- | --- | --- | --- | --- | --- |
| Seoul National University | Introduction to Bioinformatics | Midterm | Spring 2023 | Asst. Prof. M. Steinegger (martin.steinegger@snu.ac.kr) | Luna Jang (jse9512@snu.ac.kr) <br /> Rachel Kim (eunbelivable@snu.ac.kr)|

## Rules
- Using generative AI (e.g. chatGPT, copilot, etc.) is not allowed
- Open book, open note, open internet
- During exam, important questions will be answered in the [issue](https://docs.google.com/document/d/1TtI6UiyE0VopF9LteiOiGzZM9A-a7R1Tre5WXUNU30o/edit)

## Guidance
- Please write your answers to the files corresponding to the questions.<br>
    For example,
    > Question: Q1-1<br>
    > Answer: `./answer/A1-1.txt`

- Answers generated by chatGPT are provided in `gpt_answer.md` file.
  - You can use them as a reference.
  - Be careful to distinguish wrong answers generated by chatGPT.
  
- Free text questions
    - Q6-3, Q10, Q11, Q12 are free text questions.
    - Write your answers in English to corresponding text files.
    - Please keep your answers concise but containing enough information to understand.

- Submission Options (choose one)
    1. Clone the repository and submit to the repository.
    2. Edit the answer file directly at the github website.

:bangbang: Except for free text questions, all the submitted answers will be graded with
automated scripts. **Please be careful of the typos.**

## Contents  
You can go directly to the answer of each question by clicking the link below.

[Q1](#q1) [Q2](#q2) [Q3](#q3) [Q4](#q4) [Q5](#q5) [Q6](#q6) [Q7](#q7) [Q8](#q8) [Q9](#q9) [Q10](#q10) [Q11](#q11)

---

## Q1
This is the structure of `/home` folder.
```
├── harry
│   ├── chamber.txt
│   └── stone.txt
├── hermione
│   └── necklace.txt
│   └── proj2.gz
└── ron
    ├── frost.txt
    ├── proj1
    │   ├── data.txt
    │   └── workflow.sh
    └── wand.txt
```

1. Suppose you are in 'harry' directory. Write a command to move harry's chamber.txt to ron's proj1 directory, using relative path. (`A1-1.txt`) (1 point)

2. Suppose you are in ron's proj1 directory. Write a command to read first 7 lines of stone.txt. (`A1-2.txt`) (1 point)
    - Don't use 'cd'.
    - Use relative path.

3. Suppose you are in ron's proj1 directory. Move to 'hermione' directory and uncompress proj2.gz file in 'harry' directory. (`A1-3.txt`) (1 point)
    - Use 'cd' once.
    - Use relative path.

4. Using absolute path, remove every txt file in harry's directory. (`A1-4.txt`) (1 point)

---

## Q2
This is a gtf file, `Saccharomyces_cerevisiae.gtf`.

![](img/question2.png)

1. Write the expected output of this command. (`A2-1.txt`) (2 points)
    ```sh
    awk -F'\t' '{ print $3 }' Saccharomyces_cerevisiae.gtf | sort | uniq  -c
    ```

2. Write the expected output of this command. (`A2-2.txt`) (1 point)
    ```sh
    awk -F'\t' '{print NR,FNR}' Saccharomyces_cerevisiae.gtf
    ```
3. Write a command to calculate the length of the gene on the negative strand.
(`A2-3.txt`) (3 points)
> This problem can take a long time to solve.
> The GTF is provided in data directory, so use it if you need.
> You can use Google cloud shell.

4. Write result of Q2-3. (`A2-4.txt`). (1 point)

---

## Q3
This is a fasta file, `contigs.fasta`.
```
> contig1
ATGCGATCGTACGTA
CGTACGTAGCTAGCT
AGCAGTTAGGGTCGA
TCGATCGTAGCTAGC
GCTAGCTAGCTAGCT
> contig2
TAGGACCTAGTAGCG
ACAGCTCGTAGCTCT
GATACGCTCGTAGCA
CGTAGCTAGCGTAGT
GCTCGTAGCTAGCTA
```
1. Write a command to count all lines with `GTA` from `contigs.fasta`. (`A3-1.txt`) (2 points)

2. A sequence logo is a graphical representation of the sequence conservation of nucleotides. 
The relative sizes of the letters indicate their frequency in the sequences and the total height of the letters depicts the information content of the position in bits [(wikipedia)](https://en.wikipedia.org/wiki/Sequence_logo).

   This is a logo for a sequence motif.

   ![](img/question3.png)

   Write a command to find this motif and color it in `contigs.fasta`. (`A3-2.txt`) (3 points)
   > Don't consider the motifs between two lines.

3. Calculate GC content of each contig. ('A3-3.csv') (3 points)
    > Write the percent without %, and drop numbers after the decimal point.
    > Example) 50.98% and 45.12% -> **50,45**

---

## Q4
Assemble these reads using a overlap graph approach
- ACCTA
- GTACG 
- TACGT 
- GTCAC 
- CGAAC 
 

1. List all the edges of the overlap graph in `A4-1.csv` with a format as follows: (3 points) 
   > Write the edges in the order of the given reads. 

   ```
   source,target
   AGC,GCA
   ...
   ```

2. Write the final assembly result(s) (`A4-2.txt`). (4 points) 
    > Find the shortest path over the graph that visit all the nodes at least one time.
---

## Q5
1. Fill in the alignment matrix with Needleman-Wusch algorithm (`A5-1.csv`) (3 points) 
> :bangbang: You can use web resource.

type        | score
------------|------
match       | +2
mismatch    | -2
gap penalty | -4

```
| - | - | G | C | A | T | G |
|---|---|---|---|---|---|---|
| - | 0 |   |   |   |   |   |
| G |   |   |   |   |   |   |
| T |   |   |   |   |   |   |
| C |   |   |   |   |   |   |
| A |   |   |   |   |   |   |
| G |   |   |   |   |   |   |
```

2. Write the alignment result of Q5-1 (`A5-2.txt`). (1 points) 
```sh
# Alignment result format example
ATGC # Sequence of horizontal axis 
A-GC # Sequence of vertical axis
```

3. **[Free text question]** Compared to global alignment(Needleman-Wunsch algorithm), local alignmnet(Smith-Waterman algorithm) has a property that it can align to local region. What is the property? Write in one sentence. (`A5-3.txt`) (2 points)

---

## Q6

Followings are types of structural variants. 
1. Translocation
2. Deletion
3. Novel sequence insertion
4. Mobile-element insertion
5. Inversion
6. Interspersed duplication
7. Tandem duplication

<img src="img/question6.png" width="40%" height="40%">
 
Write all the types of variants occured to the query of the figure. Write each number only once and order the numbers numerically. Save the answer in csv format. (`A6.csv`) (4 points)<br>
For example, if type 1, 2, 3, 4, and 5 occured.
```
# A7.csv
1,2,3,4,5
```
:bangbang: partial credit will be given if you choose among the correct answers. If any of the wrong answers are included, no credit will be given.

---

## Q7

In SAM file format, flag score represents the property of read. The figure shows read mapped to the reference. From the figure and descriptions, calculate the flag scores of A and B. (`A7.csv`) (4 points, 2 points for each)

![](img/question7.png)<br>
> Assume the reads are mapped as pair, and both query and mate are properly mapped.<br>
> The alignments are primary and the read didn't fail for quality check.<br>
> All the reads are neither of PCR nor optical duplicates.<br>

```sh
# A7.csv
A,1 # flag score of A
B,2 # flag score of B
```
---

## Q8
1. From this multiple sequence alignment, find informative positions for drawing a phylogenetic tree. (`A8-1.csv`) (2 points)
    ```
    pos 0123456789
    s1  CGCTAGTCGA
    s2  CGGGC-TCTA
    s3  CCGTGGTCGA
    s4  CCGGT-TGTA
    ```
    > Write the indices of position (0 to 9) formatted as follows:
    ```sh
    # If answer is position 1,2, and 3, your answer should be
    1,2,3
    ```

2. Choose the tree which represents the alignment best. (`A8-2.txt`) (2 points) 
    ![](img/question8.png)
---

## Q9
**[Free text question]** 
In the context of genomic assembly, compare and contrast the effects of using affine and non-affine gap cost algorithms on the accuracy and efficiency of sequence alignment.  (`A9.txt`) (4 points) 

---

## Q10
**[Free text question]** 
Your homology search returns a top-hit with an E-value of 10. Interpret the value and what would be the next step. (`A10.txt`) (4 points)<br>
:bangbang: Write in 3 sentences.

---

## Q11
**[Free text question]** 
You are researcher who studies hyperthermophile. Using the discovered hyperthermophile genome so far, you want to disclose the evolutionary history of the species.<br>
One day, luckily you got samples of hyperthermophile. You sequenced the genome with Illumina sequencer. Now, you have reads from sequence data and want to find out if they are novel species and their evolutionary history compared to non-hyperthermophile.<br>
Using the sequenced data, please describe the workflow you are planning. The workflow have to deal with Genome assembly, Genome comparison, and Phylogenomics. You should include each stage, strategy, concepts used, software to use, etc. (`A11.txt`) (7 points)

> - There can be multiple answers. We will grade based on your understandings of workflow. 
> - You don't need to understand all the things chatGPT generated.
> - If the conditions are insufficient, you can set your own assumptions.
